### PR TITLE
Make tag_type Optional for Tag Creation

### PR DESCRIPTION
**Pull request by apekshagangurde**
Fixes issue: #132 
## Summary

Currently, creating a tag requires assigning a tag_type.

- Made the tag_type_id field optional in the tag creation model (TagCreate).
- Updated validation logic to handle the absence of tag_type_id.
- Adjusted relevant test cases to ensure tags can be created with or without a tag type.

## Summary by CodeRabbit

* **New Features**
  * Tags can now be created and updated without requiring a tag type.
  * Users can remove or assign a tag type to existing tags.
  * Bulk question uploads now support tags without explicit tag types.

* **Bug Fixes**
  * Improved error handling and messaging when assigning nonexistent or deleted tag types to tags.
  * Properly handles attempts to update or assign tag types to non-existent or deleted tags.

* **Tests**
  * Added comprehensive tests covering tag creation and updates with and without tag types, including edge cases and error scenarios.
  * Added tests for bulk uploading questions with tags lacking explicit tag types.

### CODE REVIEW

File: backend/app/tests/api/routes/test_tag.py [hunk 1]
```diff
@@ -870,3 +870,206 @@ def test_inactive_tagtype_not_listed(
...
+    get_user_superadmin_token: dict[str, str],
+) -> None:
+    user, organization = setup_user_organization(db)
+    invalid_tag_type_id = 85
+
+    data = {
+        "name": random_lower_string(),
+        "description": random_lower_string(),
+        "tag_type_id": invalid_tag_type_id,
+        "created_by_id": user.id,
```
Comment (ninad-ict) [inline backend/app/tests/api/routes/test_tag.py:938]: We dont pass created_by_ids in route payloads anymore.

File: backend/app/tests/api/routes/test_tag.py [hunk 2]
```diff
@@ -870,3 +870,206 @@ def test_inactive_tagtype_not_listed(
...
+        created_by_id=user.id,
+    )
+    db.add(tag_type)
+    db.commit()
+    db.refresh(tag_type)
+    payload = {
+        "name": random_lower_string(),
+        "description": random_lower_string(),
+        "tag_type_id": tag_type.id,
+        "created_by_id": user.id,
```
Comment (ninad-ict) [inline backend/app/tests/api/routes/test_tag.py:970]: Same comment as above for payloads

**Review by ninad-ict — CHANGES_REQUESTED** [review]
We also need to  review the Bulk uploading functionality where having a tagtype should be optional. Presently, we are tagging a tag by-default to 'training tag'

File: backend/app/tests/api/routes/test_question.py [hunk 3]
```diff
@@ -2520,3 +2520,73 @@ def test_get_questions_by_is_active_filter(
     ids = [q["id"] for q in data]
     assert inactive_question.id in ids
     assert active_question.id not in ids
+
+
+def test_bulk_upload_questions_without_tagtype(
```
Comment (ninad-ict) [inline backend/app/tests/api/routes/test_question.py:2525]: Verify if we test cases where tag types are present in bulk file and tags are created out of them

File: backend/app/api/routes/question.py [hunk 4]
```diff
@@ -92,7 +92,9 @@ def build_question_response(
             TagPublic(
                 id=tag.id,
                 name=tag.name,
-                tag_type=tag_type,
+                tag_type=get_tag_type_by_id(session, tag_type_id=tag.tag_type_id)
```
Comment (suryabulusu) [inline backend/app/api/routes/question.py:95]: get tag_type_by_id expects tag_type_id to be int.. not None

Comment (suryabulusu) [reply backend/app/api/routes/question.py:95]: i think build_question_repsonse will have issues too when tag type id is none

**Review by suryabulusu — COMMENTED** [review]
left comments

File: backend/app/api/routes/tag.py
Comment (ninad-ict) [inline backend/app/api/routes/tag.py:1]: Check the changes needed for GET Request & PATCH Request for Tag. If no tag_type, the Tag should still be returned with no tag_type

File: backend/app/api/routes/question.py [hunk 5]
```diff
@@ -1047,22 +1048,23 @@ async def upload_questions_csv(
                             tag_name = parts[1].strip()
                         else:
                             # Default tag type if no colon present
-                            tag_type_name = "Training Tag"
+                            tag_type_name = None
                             tag_name = tag_entry
 
-                        cache_key = f"{tag_type_name}:{tag_name}"
+                        cache_key = f"{tag_type_name or 'default'}:{tag_name}"
```
Comment (ninad-ict) [inline backend/app/api/routes/question.py:1054]: Can you explain why we are optionally naming as 'default' here

**Review by ninad-ict — CHANGES_REQUESTED** [review]

**Review by ninad-ict — APPROVED** [review]
